src/cpp/common/alarm.cc

```diff
@@ -52,78 +62,94 @@ class AlarmImpl : public grpc::internal::CompletionQueueTag {
     return true;
   }
   void Set(grpc::CompletionQueue* cq, gpr_timespec deadline, void* tag) {
-    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
     GRPC_CQ_INTERNAL_REF(cq->cq(), "alarm");
     cq_ = cq->cq();
     tag_ = tag;
     GPR_ASSERT(grpc_cq_begin_op(cq_, this));
-    GRPC_CLOSURE_INIT(
-        &on_alarm_,
-        [](void* arg, grpc_error_handle error) {
-          // queue the op on the completion queue
-          AlarmImpl* alarm = static_cast<AlarmImpl*>(arg);
-          alarm->Ref();
-          // Preserve the cq and reset the cq_ so that the alarm
-          // can be reset when the alarm tag is delivered.
-          grpc_completion_queue* cq = alarm->cq_;
-          alarm->cq_ = nullptr;
-          grpc_cq_end_op(
-              cq, alarm, error,
-              [](void* /*arg*/, grpc_cq_completion* /*completion*/) {}, arg,
-              &alarm->completion_);
-          GRPC_CQ_INTERNAL_UNREF(cq, "alarm");
-        },
-        this, grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&timer_,
-                    grpc_core::Timestamp::FromTimespecRoundUp(deadline),
-                    &on_alarm_);
+    Ref();
+    GPR_ASSERT(cq_armed_.exchange(true) == false);
+    GPR_ASSERT(!callback_armed_.load());
+    cq_timer_handle_ = event_engine_->RunAfter(
+        grpc_core::Timestamp::FromTimespecRoundUp(deadline) -
+            grpc_core::ExecCtx::Get()->Now(),
+        [this] {
+          cq_armed_.store(false);
```

Comment:
Is it possible to move `cq_armed_.store(false);` into `OnCQAlarm()`? Same for `callback_armed_.store(false);`.

Reply:
Good idea, done.